### PR TITLE
feat(compiler): Convert `runtime/dataStructures.gr` to primitives

### PR DESCRIPTION
**Pull request by ospencer**
This means that all of the data structures functions get inlined, so there's no performance cost to using something like `untagNumber`.

### CODE REVIEW

**Review by marcusroberts — APPROVED** [review]
From my limited understanding looks good. I see you've updated the formatter, the new LSP may need updating when merged too, maybe not if oprint is doing the work here.

**Review by peblair — APPROVED** [review]
I love it.

67kced

File: compiler/src/codegen/compcore.re [hunk 1]
```diff
@@ -277,80 +220,12 @@ let grain_function_imports = [
...
     mimp_mod: equal_mod,
     mimp_name: equal_ident,
     mimp_type:
       MFuncImport(
-        [
-          Types.StackAllocated(WasmI32),
-          Types.StackAllocated(WasmI32),
-          Types.StackAllocated(WasmI32),
-        ],
+        [Types.HeapAllocated, Types.HeapAllocated, Types.HeapAllocated],
```
Comment (phated) [inline compiler/src/codegen/compcore.re:228]: Why is this changed to heap allocated?

File: stdlib/runtime/dataStructures.gr [hunk 2]
```diff
@@ -1,350 +1,176 @@
...
+export primitive allocateChar: () -> WasmI32 = "@allocate.char"
 
 // INT32/INT64
 
 /**
- * Allocates a new Int64.
+ * Allocates a new Int32.
  *
- * @returns {WasmI32}
+ * @returns
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:52]: Gave up on writing these?

File: stdlib/runtime/dataStructures.gr [hunk 3]
```diff
@@ -1,350 +1,176 @@
...
-
-  ptr
-}
+@unsafe
+export primitive allocateInt32: () -> WasmI32 = "@allocate.int32"
 
 /**
- * Allocates a new Int64 with a prepopulated value
+ * Allocates a new Int32 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:59]: No `@returns`?

File: stdlib/runtime/dataStructures.gr [hunk 4]
```diff
@@ -1,350 +1,176 @@
...
-export let loadInt64 = xptr => {
-  WasmI64.load(xptr, 8n)
-}
-
-/**
- * Allocates a new Int32.
+ * Allocates a new Int64.
  *
- * @returns {WasmI32}
+ * @returns
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:67]: Gave up on writing these?

File: stdlib/runtime/dataStructures.gr [hunk 5]
```diff
@@ -1,350 +1,176 @@
...
-
-  ptr
-}
+@unsafe
+export primitive allocateInt64: () -> WasmI32 = "@allocate.int64"
 
 /**
- * Allocates a new Int32 with a prepopulated value
+ * Allocates a new Int64 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:74]: No `@returns`?

File: stdlib/runtime/dataStructures.gr [hunk 6]
```diff
@@ -1,350 +1,176 @@
...
+@unsafe
+export primitive newInt64: WasmI64 -> WasmI32 = "@new.int64"
 
 // FLOATS
 
 /**
  * Allocates a new Float32.
  *
- * @returns {WasmI32}
+ * @returns
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:84]: Gave up on writing these?

File: stdlib/runtime/dataStructures.gr [hunk 7]
```diff
@@ -1,350 +1,176 @@
...
-  WasmI32.store(ptr, Tags._GRAIN_FLOAT32_BOXED_NUM_TAG, 4n)
-
-  ptr
-}
+@unsafe
+export primitive allocateFloat32: () -> WasmI32 = "@allocate.float32"
 
 /**
  * Allocates a new Float32 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:91]: No `@returns`?

File: stdlib/runtime/dataStructures.gr [hunk 8]
```diff
@@ -1,350 +1,176 @@
...
-  WasmF32.load(xptr, 8n)
-}
+@unsafe
+export primitive newFloat32: WasmF32 -> WasmI32 = "@new.float32"
 
 /**
  * Allocates a new Float64.
  *
- * @returns {WasmI32}
+ * @returns
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:99]: Gave up on writing these?

File: stdlib/runtime/dataStructures.gr [hunk 9]
```diff
@@ -1,350 +1,176 @@
...
-  WasmI32.store(ptr, Tags._GRAIN_FLOAT64_BOXED_NUM_TAG, 4n)
-
-  ptr
-}
+@unsafe
+export primitive allocateFloat64: () -> WasmI32 = "@allocate.float64"
 
 /**
  * Allocates a new Float64 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:106]: No `@returns`?

File: stdlib/runtime/dataStructures.gr [hunk 10]
```diff
@@ -1,350 +1,176 @@
...
+@unsafe
+export primitive newFloat64: WasmF64 -> WasmI32 = "@new.float64"
 
 // RATIONALS
 
 /**
  * Allocates a new Rational.
  *
- * @returns {WasmI32}
+ * @returns
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:116]: Gave up on writing these?

File: stdlib/runtime/dataStructures.gr [hunk 11]
```diff
@@ -1,350 +1,176 @@
...
-  WasmI32.store(ptr, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG, 4n)
-
-  ptr
-}
+@unsafe
+export primitive allocateRational: () -> WasmI32 = "@allocate.rational"
 
 /**
  * Allocates a new Rational with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:123]: Wrong docs? Can you please update

File: stdlib/runtime/dataStructures.gr [hunk 12]
```diff
@@ -1,350 +1,176 @@
...
-}
+@unsafe
+export primitive newRational: (WasmI32, WasmI32) -> WasmI32 = "@new.rational"
 
 /**
- * Returns a pointer to the heap location containing this boxed number's Rational numerator
- * @param wrappedRational The boxed rational to return
+ * Load the (tagged) variant of an ADT.
+ *
+ * @export
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:131]: wat

File: stdlib/runtime/dataStructures.gr [hunk 13]
```diff
@@ -1,350 +1,176 @@
...
-}
+@unsafe
+export primitive loadAdtVariant: WasmI32 -> WasmI32 = "@adt.load_variant"
 
 /**
- * Returns a pointer to the heap location containing this boxed number's Rational numerator
- * @param wrappedRational The boxed rational to return
+ * Load an untagged string's size.
+ *
+ * @export
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:141]: wat?

File: stdlib/runtime/dataStructures.gr [hunk 14]
```diff
@@ -1,350 +1,176 @@
...
-  WasmI32.load(xptr, 12n)
-}
+@unsafe
+export primitive stringSize: WasmI32 -> WasmI32 = "@string.size"
 
 /**
- * Load a value from an ADT.
+ * Load an untagged Bytes' size.
  *
  * @export
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:151]: wat?

File: stdlib/runtime/dataStructures.gr [hunk 15]
```diff
@@ -1,350 +1,176 @@
...
-  WasmI32.load(ptr, 12n)
-}
+@unsafe
+export primitive tagSimpleNumber: WasmI32 -> Number = "@tag.simple_number"
 
 /**
- * Load an untagged string's size.
+ * Untag a simple number.
  *
  * @export
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:171]: wat?

**Review by phated — CHANGES_REQUESTED** [review]
Since you are doing a lot of work in dataStructures, can you please make sure the graindoc is correct throughout?

File: compiler/src/codegen/compcore.re (hunk 1 above)
Comment (ospencer) [reply compiler/src/codegen/compcore.re:228]: Was just wrong before. This is a little bit of a subtle bug to catch too, since it'd only be a problem if you were pattern matching on heap-allocated constants. I only noticed it because of the refactor.

File: stdlib/runtime/dataStructures.gr [hunk 16]
```diff
@@ -1,350 +1,177 @@
...
-  arr
-}
-
-/**
- * Stores an item in a Grain array.
- *
- * @param {WasmI32} array The array to store the item in
- * @param {WasmI32} idx The index to store the item
- * @param {WasmI32} item The item to store
+ * @param numElts The number of elements to be contained in this array
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:6]:
```suggestion
 * @param numElts: The number of elements to be contained in this array
```

File: stdlib/runtime/dataStructures.gr [hunk 17]
```diff
@@ -1,350 +1,177 @@
...
-}
+@unsafe
+export primitive allocateArray: WasmI32 -> WasmI32 = "@allocate.array"
 
 /**
  * Allocates a new Grain tuple.
  *
- * @param {WasmI32} numElts The number of elements to be contained in this tuple
- * @returns {WasmI32} The pointer to the tuple
+ * @param numElts The number of elements to be contained in this tuple
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:15]:
```suggestion
 * @param numElts: The number of elements to be contained in this tuple
```

File: stdlib/runtime/dataStructures.gr [hunk 18]
```diff
@@ -1,350 +1,177 @@
...
-}
+@unsafe
+export primitive allocateTuple: WasmI32 -> WasmI32 = "@allocate.tuple"
 
 /**
  * Allocates a new Grain bytes.
  *
- * @param {WasmI32} size The number of bytes to be contained in this buffer
- * @returns {WasmI32} The pointer to the bytes
+ * @param size The number of bytes to be contained in this buffer
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:24]:
```suggestion
 * @param size: The number of bytes to be contained in this buffer
```

File: stdlib/runtime/dataStructures.gr [hunk 19]
```diff
@@ -1,350 +1,177 @@
...
-}
+@unsafe
+export primitive allocateBytes: WasmI32 -> WasmI32 = "@allocate.bytes"
 
 /**
  * Allocates a new Grain string.
  *
- * @param {WasmI32} size The size (in bytes) of the string to allocate
- * @returns {WasmI32} The pointer to the string
+ * @param size The size (in bytes) of the string to allocate
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:33]:
```suggestion
 * @param size: The size (in bytes) of the string to allocate
```

File: stdlib/runtime/dataStructures.gr [hunk 20]
```diff
@@ -1,350 +1,177 @@
...
-
-  ptr
-}
+@unsafe
+export primitive allocateInt32: () -> WasmI32 = "@allocate.int32"
 
 /**
- * Allocates a new Int64 with a prepopulated value
+ * Allocates a new Int32 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:59]:
```suggestion
 * @param value: The value to store
```

File: stdlib/runtime/dataStructures.gr [hunk 21]
```diff
@@ -1,350 +1,177 @@
...
-
-  ptr
-}
+@unsafe
+export primitive allocateInt64: () -> WasmI32 = "@allocate.int64"
 
 /**
- * Allocates a new Int32 with a prepopulated value
+ * Allocates a new Int64 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:75]:
```suggestion
 * @param value: The value to store
```

File: stdlib/runtime/dataStructures.gr [hunk 22]
```diff
@@ -1,350 +1,177 @@
...
-  WasmI32.store(ptr, Tags._GRAIN_FLOAT32_BOXED_NUM_TAG, 4n)
-
-  ptr
-}
+@unsafe
+export primitive allocateFloat32: () -> WasmI32 = "@allocate.float32"
 
 /**
  * Allocates a new Float32 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:93]:
```suggestion
 * @param value: The value to store
```

File: stdlib/runtime/dataStructures.gr [hunk 23]
```diff
@@ -1,350 +1,177 @@
...
-  WasmI32.store(ptr, Tags._GRAIN_FLOAT64_BOXED_NUM_TAG, 4n)
-
-  ptr
-}
+@unsafe
+export primitive allocateFloat64: () -> WasmI32 = "@allocate.float64"
 
 /**
  * Allocates a new Float64 with a prepopulated value
  * @param value The value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:109]:
```suggestion
 * @param value: The value to store
```

File: stdlib/runtime/dataStructures.gr [hunk 24]
```diff
@@ -1,350 +1,177 @@
...
-  ptr
-}
+@unsafe
+export primitive allocateRational: () -> WasmI32 = "@allocate.rational"
 
 /**
  * Allocates a new Rational with a prepopulated value
- * @param value The value to store
+ * @param value The numerator value to store
+ * @param value The denominator value to store
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:128]:
```suggestion
 * @param value: The numerator value to store
 * @param value: The denominator value to store
```

File: stdlib/runtime/dataStructures.gr [hunk 25]
```diff
@@ -1,350 +1,177 @@
...
-}
+@unsafe
+export primitive newRational: (WasmI32, WasmI32) -> WasmI32 = "@new.rational"
 
 /**
- * Returns a pointer to the heap location containing this boxed number's Rational numerator
- * @param wrappedRational The boxed rational to return
+ * Load the (tagged) variant of an ADT.
+ *
+ * @param ptr Untagged pointer to the ADT
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:137]:
```suggestion
 * @param ptr: Untagged pointer to the ADT
```

File: stdlib/runtime/dataStructures.gr [hunk 26]
```diff
@@ -1,350 +1,177 @@
...
-}
+@unsafe
+export primitive loadAdtVariant: WasmI32 -> WasmI32 = "@adt.load_variant"
 
 /**
- * Returns a pointer to the heap location containing this boxed number's Rational numerator
- * @param wrappedRational The boxed rational to return
+ * Load an untagged string's size.
+ *
+ * @param ptr Untagged pointer to the string
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:146]:
```suggestion
 * @param ptr: Untagged pointer to the string
```

File: stdlib/runtime/dataStructures.gr [hunk 27]
```diff
@@ -1,350 +1,177 @@
...
 
 /**
- * Load a value from an ADT.
+ * Load an untagged Bytes' size.
  *
- * @export
- * @param {WasmI32} ptr Untagged pointer to the ADT
- * @param {WasmI32} idx Index (from zero) of the item
- * @returns {WasmI32} The value located at the index
+ * @param ptr Untagged pointer to the Bytes
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:155]:
```suggestion
 * @param ptr: Untagged pointer to the Bytes
```

File: stdlib/runtime/dataStructures.gr [hunk 28]
```diff
@@ -1,350 +1,177 @@
...
+export primitive bytesSize: WasmI32 -> WasmI32 = "@bytes.size"
 
 /**
- * Load the (tagged) variant of an ADT.
+ * Tag a simple number.
  *
- * @export
- * @param {WasmI32} ptr Untagged pointer to the ADT
- * @returns {WasmI32} The (tagged) ADT variant id
+ * @param num The number to tag
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:164]:
```suggestion
 * @param num: The number to tag
```

File: stdlib/runtime/dataStructures.gr [hunk 29]
```diff
@@ -1,350 +1,177 @@
...
+export primitive tagSimpleNumber: WasmI32 -> Number = "@tag.simple_number"
 
 /**
- * Load an untagged string's size.
+ * Untag a simple number.
  *
- * @export
- * @param {WasmI32} ptr Untagged pointer to the string
- * @returns {WasmI32} The string size (in bytes)
+ * @param num The number to untag
```
Comment (phated) [inline stdlib/runtime/dataStructures.gr:173]:
```suggestion
 * @param num: The number to untag
```

**Review by phated — CHANGES_REQUESTED** [review]
Slight mistake on the graindocs. Can you generate the markdown file for this file so we can know that everything is doc'd correctly?

**Comment by ospencer** [comment]
@phated Let me know if that looks good.

**Review by phated — APPROVED** [review]
Good enough! I had some stylistic nits but not worth it for this internal documentation.
